src/nanoemoji/write_font.py

```diff
@@ -296,35 +366,44 @@ def _draw_glyph_extents(
     return glyph
 
 
-def _glyf_ufo(ufo, color_glyphs):
+def _glyf_ufo(
+    config: FontConfig, ufo: ufoLib2.Font, color_glyphs: MutableSequence[ColorGlyph]
+):
     # glyphs by reuse_key
-    glyphs = {}
-    reused = set()
-    for color_glyph in color_glyphs:
+    glyph_cache = GlyphReuseCache(config)
+    glyph_uses = defaultdict(int)
+    for i, color_glyph in enumerate(color_glyphs):
         logging.debug(
             "%s %s %s",
             ufo.info.familyName,
             color_glyph.glyph_name,
             color_glyph.transform_for_font_space(),
         )
         parent_glyph = ufo[color_glyph.glyph_name]
-        for painted_layer in color_glyph.painted_layers:
-            # if we've seen this shape before reuse it
-            reuse_key = painted_layer.shape_cache_key()
-            if reuse_key not in glyphs:
-                glyph = _create_glyph(color_glyph, painted_layer)
-                glyphs[reuse_key] = glyph
-            else:
-                glyph = glyphs[reuse_key]
-                reused.add(glyph.name)
-            parent_glyph.components.append(Component(baseGlyph=glyph.name))
 
+        # generate glyphs for PaintGlyph's and assign glyph names
+        color_glyphs[i] = color_glyph = _migrate_paths_to_ufo_glyphs(
+            color_glyph, glyph_cache
+        )
+
+        for root in color_glyph.painted_layers:
+            for context in root.breadth_first():
+                # For 'glyf' just dump anything that isn't a PaintGlyph
```

Comment:
should we log a warning if that isn't the case?

Reply:
I can ... concern was there isn't really anything to be done about it. `glyf` just can't produce all the things COLR can.